### PR TITLE
[CGPROD-2706] item config auto loading

### PR DESCRIPTION
**Pull request by BeeGodwin**
- Renamed 'registry' to 'catalogue'
- Loader now loads catalogue
- configs can specify a catalogueKey: if they do, loader looks for a matching json5 in `themes/default/items`
- loader brings those json5s into the cache and uses them to initialise the catalogue as per CGPROD-2705
- updated docs, added tests.

### CODE REVIEW

File: src/core/loader/get-registry.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    });
+};
+
+const getRegistryKeys = config => {
+    const registryKey = item => item[1].registryKey;
+
+    return Object.entries(config).filter(registryKey).map(registryKey);
+};
+
+const loadToCache = (screen, key) => {
```
Comment (FostUK) [inline src/core/loader/get-registry.js:32]: I think, if you make this **const loadToCache = screen => key => {**

then you can do 

**registryKeys.forEach(loadToCache(screen);**

on line 13 above. Might be worth us changing that in the other bit of loader at some point...

File: src/core/screen.js [hunk 2]
```diff
@@ -45,6 +45,7 @@ export class Screen extends Phaser.Scene {
             activeScreens,
             navigation: this._data.navigation,
             transientData: this._data.transient || {},
+            registry: this._data.registry,
```
Comment (FostUK) [inline src/core/screen.js:48]: As discussed - this isn't required. In generally we also avoid sharing anything as much as possible via classes what with classes being terrible - unfortunately Phaser being class based makes this somewhat difficult to avoid.

File: src/core/loader/get-registry.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
+/**
+ * @copyright BBC 2020
+ * @author BBC Children's D+E
+ * @license Apache-2.0
+ */
+import { itemsRegistry, initRegistry } from "../../components/shop/item-registry.js";
+
+export const loadRegistry = (screen, config) => {
+    const registryKeys = getRegistryKeys(config);
+    if (!registryKeys.length) return;
```
Comment (FostUK) [inline src/core/loader/get-registry.js:10]: If there are no items in the config will it return an empty array? Prob a good pattern to do that.

You can delete this line if so.

Comment (BeeGodwin) [reply src/core/loader/get-registry.js:10]: It will, and you're right that the line isn't strictly needed, but early return avoids the need to modify the rest of the loader tests plus I thought it had semantic value. Happy to delete on the 'less is more' principle if you prefer.

File: src/core/screen.js (hunk 2 above)
Comment (BeeGodwin) [reply src/core/screen.js:48]: Yeah, meant to take this out. TY =)

File: src/core/loader/get-registry.js (hunk 1 above)
Comment (BeeGodwin) [reply src/core/loader/get-registry.js:32]: I don't think I've seen (or, at least, understood) that double arrow pattern, but it's currying isn't it? Lovely little bit of functional goodness!

File: src/core/loader/get-registry.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+            initRegistry(key, screen.cache.json.get(`registry-${key}`));
+        });
+        console.log("BEEBUG: itemsRegistry", itemsRegistry);
+    });
+};
+
+const getRegistryKeys = config => {
+    const registryKey = item => item[1].registryKey;
+
+    return Object.entries(config).filter(registryKey).map(registryKey);
```
Comment (FostUK) [inline src/core/loader/get-registry.js:29]: Is the filter trying to remove undefined entries?
Miht be clearer to map first then .filter(Boolean) ?

File: src/core/loader/get-registry.js [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    const registryKeys = getRegistryKeys(config);
+    if (!registryKeys.length) return;
+
+    registryKeys.forEach(key => {
+        loadToCache(screen, key);
+    });
+
+    screen.load.start();
+
+    screen.load.on("complete", () => {
```
Comment (AdamXA) [inline src/core/loader/get-registry.js:18]: define complete callback before starting loader

File: src/core/loader/get-registry.js [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+ */
+import { itemsRegistry, initRegistry } from "../../components/shop/item-registry.js";
+
+export const loadRegistry = (screen, config) => {
+    const registryKeys = getRegistryKeys(config);
+    if (!registryKeys.length) return;
+
+    registryKeys.forEach(key => {
+        loadToCache(screen, key);
+    });
```
Comment (AdamXA) [inline src/core/loader/get-registry.js:14]: can be done in one line `registryKeys.forEach(key => loadToCache(screen, key));`

File: src/core/loader/get-registry.js (hunk 1 above)
Comment (BeeGodwin) [reply src/core/loader/get-registry.js]: 672e4a9

File: src/core/screen.js (hunk 2 above)
Comment (BeeGodwin) [reply src/core/screen.js]: 672e4a9

File: src/core/loader/get-registry.js (hunk 6 above)
Comment (BeeGodwin) [reply src/core/loader/get-registry.js]: Used the currying approach suggested by Nick in the comment above, per 672e4a9

File: src/core/loader/get-registry.js (hunk 4 above)
Comment (BeeGodwin) [reply src/core/loader/get-registry.js]: I'm learning all _sorts_ of functional JS tricks today. TY!

File: src/core/loader/get-registry.js (hunk 3 above)
Comment (BeeGodwin) [reply src/core/loader/get-registry.js]: On reflection, leaving it in is a code smell- it's better to make the needed changes in the loader tests (just adding a mock anyway.)

File: src/core/loader/get-catalogue.js [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+ * @license Apache-2.0
+ */
+import { initCatalogue } from "../../components/shop/item-catalogue.js";
+
+export const loadCatalogue = (screen, config) => {
+    const catalogueKeys = getCatalogueKeys(config);
+
+    catalogueKeys.forEach(loadToCache(screen));
+
+    screen.load.on("complete", () => {
```
Comment (AdamXA) [inline src/core/loader/get-catalogue.js:13]: this only needs to be called once I guess, so you can use `.once` instead

Comment (AdamXA) [reply src/core/loader/get-catalogue.js:13]: could probably get rid of some of the `{ }` in this callback too

Comment (BeeGodwin) [reply src/core/loader/get-catalogue.js]: f84677e

File: docs/development/shop-component/items.md [hunk 8]
```diff
@@ -1,27 +1,39 @@
...
-import { initRegistry } from "../core/item-registry.js";
+Your scene config should contain a `catalogueKey` which should be a string corresponding to the name of a .json5 file in `themes/default/items`. The loader will then pick them up and make them available as a map. 
+ 
 
-const registry = initRegistry("registryKey", ['array' , 'of', 'items'])); // tbd
+```javascript
+import { catalogue } from "./components/shop/item-catalogue.js";
+const catalogueSection = catalogue.get(sceneConfig.catalogueKey); // corresponding to a .json5 in items/
+const snotulonCannon = catalogueSection.get("Snotulon Cannon"); // get a single item
+const weapons = catalogueSection.getCategory(["weapons"s]); // filter by array of categories
```
Comment (AdamXA) [inline docs/development/shop-component/items.md:25]: typo

File: src/core/loader/get-catalogue.js [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+        catalogueKeys.forEach(key => initCatalogue(key, screen.cache.json.get(`catalogue-${key}`))),
+    );
+
+    screen.load.start();
+};
+
+export const getCatalogueKeys = config => {
+    const catalogueKey = item => item[1].catalogueKey;
+
+    return Object.entries(config).map(catalogueKey).filter(Boolean);
```
Comment (AdamXA) [inline src/core/loader/get-catalogue.js:23]: can you just use `Object.values` instead of using entries to get the value? (finding it hard to follow what's happening here)

Comment (AdamXA) [reply src/core/loader/get-catalogue.js:23]: I'm guessing it could just end up as `Object.values(config).filter(Boolean);`

File: docs/development/shop-component/items.md (hunk 8 above)
Comment (AdamXA) [reply docs/development/shop-component/items.md:25]: "weapon"s

File: src/core/loader/get-catalogue.js (hunk 9 above)
Comment (BeeGodwin) [reply src/core/loader/get-catalogue.js]: 5a44dfc

File: src/components/shop/item-catalogue.js [hunk 10]
```diff
@@ -26,14 +26,14 @@ export const initRegistry = (key, items) => {
         return false;
     };
 
-    const itemList = {
+    const catalogueSection = {
```
Comment (FostUK) [inline src/components/shop/item-catalogue.js:29]: Prefer itemList for this. Does what it says on the tin and this is the end agencies will deal with.

File: src/components/shop/item-catalogue.js [hunk 11]
```diff
@@ -5,11 +5,11 @@
  */
 import fp from "../../../lib/lodash/fp/fp.js";
```
Comment (FostUK) [inline src/components/shop/item-catalogue.js:6]: Is this used still?

Comment (BeeGodwin) [reply src/components/shop/item-catalogue.js:6]: Yep, currently on line 23.

File: src/components/shop/item-catalogue.js (hunk 10 above)
Comment (BeeGodwin) [reply src/components/shop/item-catalogue.js]: 9b5c1cb

**Review by AnthonyGoodchild — APPROVED** [review]
lgtm 🛒
